File: Middle of the Linked List - Leetcode 876/Middle of the Linked List - Leetcode 876.ts
class ListNode {
  public val: number;
  public next: ListNode;
  constructor(val: number, next: ListNode) {
    this.val = val;
    this.next = next;
  }
}
// Solution 1: T - O(n), S - (1)
const middleLinkedList = (head: ListNode) => {
  let length = 0;
  let c = head;
  while (c) {
    length++;
    c = c.next;
  }
  let cnt = 0;
  while (cnt < Math.floor(length)) {
    head = head.next;
  }

  return head;
};

// Solution 2: T - O(n), S - (1)
// with Two pointer slow and fast

const middleLinkedList2P = (head: ListNode) => {
  let s = head,
    f = head;

  while (f && f.next) {
    f = f.next.next;
    s = s.next;
  }

  return s;
};

File: Middle of the Linked List - Leetcode 876/Middle of the Linked List - Leetcode 876.java
class ListNode {
    int val;
    ListNode next;

    ListNode(){}

    ListNode(int val){
        this.val = val;
    }
    ListNode(int val, ListNode next){
        this.val = val;
        this.next = next;
    }
}

class Solution {
    public ListNode middleLinkedList(ListNode head){
        int length = 0;
        ListNode c = head;

        while (c!=null)
        {
            length++;
            c = c.next;
        }

        int cnt = 0;

        while(cnt < length/2)
        {
            head = head.next;
            cnt++;
        }

        return head;
    }

    public ListNode middleLinkedList2P(ListNode head){
        ListNode s = head;
        ListNode f = head;

        while (f && f.next){
            f = f.next.next;
            s = s.next;
        }

        return s;
    }
}